File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        visited = set()
        count = 0
        for r in range(len(grid)):
            for c in range(len(grid[0])):
               if self.explore(grid, r, c, visited):
                    count += 1
        return count
    
    def explore(self, grid, r, c, visited):
        # base cases
        rowInbounds = 0 <= r < len(grid)
        colInbounds = 0 <= c < len(grid[0])
        if not rowInbounds or not colInbounds:
            return False
        if (r,c) in visited:
            return False
        if grid[r][c] == "0":
            return False
        visited.add((r,c))

        self.explore(grid, r+1, c, visited)
        self.explore(grid, r-1, c, visited)
        self.explore(grid, r, c+1, visited)
        self.explore(grid, r, c-1, visited)

        return True

File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        Set<String> visited = new HashSet<>();
        int count = 0;
        for (int i=0; i<grid.length; i++) {
            for (int j=0; j<grid[0].length; j++) {
                if (dfs(i, j, visited, grid)) {
                    count += 1;
                }
            }
        }
        return count;
    }

    public boolean dfs(int r, int c, Set<String> visited, char[][] grid) {
        if (r < 0 || r >= grid.length ||
         c < 0 || c >= grid[0].length || grid[r][c] == '0' || 
        visited.contains(r + "+" + c)) {
            return false;
        }
        visited.add(r + "+" + c);
        dfs(r+1, c, visited, grid);
        dfs(r-1, c, visited, grid);
        dfs(r, c+1, visited, grid);
        dfs(r, c-1, visited, grid);

        return true;
    }
}